From 8b1a40d7a4f6dbd7c100aff6acee7b393ce36ab0 Mon Sep 17 00:00:00 2001
From: Li Jun <jun.li@freescale.com>
Date: Thu, 2 Jul 2015 08:13:39 +0800
Subject: [PATCH 0153/5242] MLK-11183 usb: chipidea: otg: complete state
 machine before cancel hrtimer

commit  39aa19578ea58fd46cf873028e2368ec2926fdc5 from
https://source.codeaurora.org/external/imx/linux-imx.git

When do usb driver module load and unload loop test for A-device, the
hrtimer may be cancelled during otg state machine running, so after
the hrtimer is cancelled, the otg state machine may start it again by
adding a new timer, this will cause the hrtimer will be active after
its function memory is freed by module unload. This patch is to fix it
by trying to hold the fsm mutex lock before setting otg state to be
undefined, then otg hrtimer will be cancelled after current otg fsm
transition completes, and no new timer will be added.

Signed-off-by: Li Jun <jun.li@freescale.com>
(cherry picked from commit 8c3ab4f22bc34675aeb8f9c62e69b450a179be0f)
(cherry picked from commit 95c8590c5232eeb9d054ac7504455a90a133f05a)
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/usb/chipidea/otg_fsm.c |    5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/drivers/usb/chipidea/otg_fsm.c b/drivers/usb/chipidea/otg_fsm.c
index 1249925..6dd38da 100644
--- a/drivers/usb/chipidea/otg_fsm.c
+++ b/drivers/usb/chipidea/otg_fsm.c
@@ -932,6 +932,10 @@ void ci_hdrc_otg_fsm_remove(struct ci_hdrc *ci)
 {
 	enum otg_fsm_timer i;
 
+	mutex_lock(&ci->fsm.lock);
+	ci->fsm.otg->state = OTG_STATE_UNDEFINED;
+	mutex_unlock(&ci->fsm.lock);
+
 	for (i = 0; i < NUM_OTG_FSM_TIMERS; i++)
 		otg_del_timer(&ci->fsm, i);
 
@@ -942,7 +946,6 @@ void ci_hdrc_otg_fsm_remove(struct ci_hdrc *ci)
 		otg_drv_vbus(&ci->fsm, 0);
 
 	sysfs_remove_group(&ci->dev->kobj, &inputs_attr_group);
-	ci->fsm.otg->state = OTG_STATE_UNDEFINED;
 }
 
 /* Restart OTG fsm if resume from power lost */
-- 
1.7.9.5

